homeassistant/components/nasweb/coordinator.py

```diff
@@ -0,0 +1,175 @@
+        if nasweb_coordinator is None:
+            _LOGGER.warning("Received notification for not registered nasweb")
+            return None
+        await nasweb_coordinator.handle_push_notification(notification)
+        return Response(body='{"response": "ok"}', content_type="application/json")
+
+
+class NASwebCoordinator(BaseDataUpdateCoordinatorProtocol):
+    """Coordinator managing status of single NASweb device."""
+
+    def __init__(
+        self, hass: HomeAssistant, webio_api: WebioAPI, name: str = "NASweb[default]"
+    ) -> None:
+        """Initialize NASweb coordinator."""
+        self._hass = hass
+        self.name = name
+        self.webio_api: WebioAPI = webio_api
+        self._last_update: float | None = None
+        job_name = f"NASwebCoordinator[{name}]"
+        self._job = HassJob(self._handle_max_update_interval, job_name)
+        self._unsub_last_update_check: CALLBACK_TYPE | None = None
+        self._listeners: dict[CALLBACK_TYPE, tuple[CALLBACK_TYPE, object | None]] = {}
+        data: dict[str, Any] = {}
+        data[KEY_OUTPUTS] = self.webio_api.outputs
+        self.async_set_updated_data(data)
+
+    def is_connection_confirmed(self) -> bool:
+        """Check whether coordinator received status update from NASweb."""
+        return self._last_update is not None
+
+    @callback
+    def async_add_listener(
+        self, update_callback: CALLBACK_TYPE, context: Any = None
+    ) -> Callable[[], None]:
+        """Listen for data updates."""
+        schedule_update_check = not self._listeners
+
+        @callback
+        def remove_listener() -> None:
+            """Remove update listener."""
+            self._listeners.pop(remove_listener)
+            if not self._listeners:
+                self._async_unsub_last_update_check()
+
+        self._listeners[remove_listener] = (update_callback, context)
+        # This is the first listener, set up interval.
+        if schedule_update_check:
+            self._schedule_last_update_check()
+        return remove_listener
+
+    @callback
+    def async_set_updated_data(self, data: dict[str, Any]) -> None:
+        """Update data and notify listeners."""
+        self.data = data
+        self.last_update = self._hass.loop.time()
+        _LOGGER.debug("Updated %s data", self.name)
+        if self._listeners:
+            self._schedule_last_update_check()
+        self.async_update_listeners()
+
+    @callback
+    def async_update_listeners(self) -> None:
+        """Update all registered listeners."""
+        for update_callback, _ in list(self._listeners.values()):
+            update_callback()
+
+    async def _handle_max_update_interval(self, now: datetime) -> None:
+        """Handle max update interval occurrence."""
+        self._unsub_last_update_check = None
+        if self._listeners:
+            self.async_update_listeners()
+
+    def _schedule_last_update_check(self) -> None:
```

Comment:
Can you explain how the methods `_schedule_last_update_check` and `_handle_max_update_interval` work?

Reply:
Since status updates are handled exclusively by push notifications, I needed a way to determine that entities had not been updated for an extended period of time. `_schedule_last_update_check` adds a task (`_handle_max_update_interval`) to be executed after `STATUS_UPDATE_MAX_TIME_INTERVAL`. This task is rescheduled whenever status updates are received. If the task is actually executed, it updates entities, which enables them to change their status to unavailable (as evidenced by the last status update being too old). The device sends status updates when there is a status change or after a set interval (in the case of no status changes).

Reply:
Oh, that makes sense!

Please explain this in comments in the code. Maybe add it to the docstring of `NASwebCoordinator`